src/basic/build-path.c

```diff
@@ -0,0 +1,232 @@
+                        return get_runpath_from_dynamic((const ElfW(Dyn)*) p->p_vaddr, ret);
+        }
+
+        return -ENOTRECOVERABLE;
+}
+
+int get_build_exec_dir(char **ret) {
+        int r;
+
+        /* Returns the build execution directory if we are invoked in a build environment. Specifically, this
+         * checks if the main program binary has an rpath/runpath set (i.e. an explicit directory where to
+         * look for shared libraries). If so we know that this is not a regular installed binary, but one
+         * which shall acquire its libraries from some specific directory, i.e. a build directory or
+         * similar. In that case it typically makes sense to also search for our auxiliary executables we
+         * fork() off in a directory close to our main program binary, rather than in the system.
+         *
+         * This function is supposed to be used when looking for "callout" binaries that are closely related
+         * to the main program (i.e. speak a specific protocol between each other). And where it's generally
+         * a good idea to use the binary from the build tree (if there is one) instead of the system.
+         *
+         * Note that this does *not* actually return the rpath/runpath but the instead the directory the main
+         * executable was found in. This follows the logic that the result is supposed to be used for
+         * executable binaries (i.e. stuff in bindir), not for shared libraries (i.e. stuff in libdir), and
+         * hence the literal shared library path would just be wrong.
+         *
+         * TLDR: if we look for callouts in this dir first, running binaries from the meson build tree
+         * automatically uses the right callout.
+         *
+         * Returns:
+         *     -ENOEXEC         → We are not running in an rpath/runpath environment
+         *     -ENOENT          → We don't know our own binary path
+         *     -NOTRECOVERABLE  → Dynamic binary information missing?
+         */
+
+        static int runpath_cached = -ERRNO_MAX-1;
+        if (runpath_cached == -ERRNO_MAX-1)
+                runpath_cached = get_runpath(NULL);
+        if (runpath_cached < 0)
+                return runpath_cached;
+        if (runpath_cached == 0)
+                return -ENOEXEC;
+
+        _cleanup_free_ char *exe = NULL;
+        r = get_process_exe(0, &exe);
+        if (r < 0)
+                return runpath_cached = r;
+
+        return path_extract_directory(exe, ret);
+}
+
+static int find_build_dir_binary(const char *path, char **ret) {
+        int r;
+
+        assert(path);
+        assert(ret);
+
+        _cleanup_free_ char *build_dir = NULL;
+        r = get_build_exec_dir(&build_dir);
+        if (r < 0)
+                return r;
+
+        _cleanup_free_ char *fn = NULL;
+        r = path_extract_filename(path, &fn);
+        if (r < 0)
+                return r;
+        if (r == O_DIRECTORY) /* Uh? */
+                return -EISDIR;
```

Comment:
i moved this now, but I left the EISDIR in it. While right now we just pass a hardcoded path in each case I think there's value in keeping this somewhat generic, and in allow possibly user-controlled paths for this too. Or to say this differently: if this was "main program" code I think an assert() would make a ton of sense, since really know how we are calling this. But given this is "library-like" code, I think we should keep a modicum of flexibility here, and rather refuse bad input politely than assert on it brutally. If that makes any sense?